File: python/0002_add_two_numbers.py
# https://leetcode.com/problems/add-two-numbers/
# O(max(n, m)) time complexity, where n and m are length of respective linked lists
# O(1) space complexity
def add_two_numbers(l1, l2):
    num1 = num2 = i = j = 0
    while l1 is not None:
        num1 += l1.val * 10**i
        i += 1
        l1 = l1.next
    while l2 is not None:
        num2 += l2.val * 10**j
        j += 1
        l2 = l2.next
    sum = num1 + num2
    dummy = curr = ListNode()
    while sum > 0:
        curr.next = ListNode()
        curr = curr.next
        curr.val = sum % 10
        sum //= 10
    return dummy.next


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

File: javascript/0002_add_two_numbers.js
// https://leetcode.com/problems/add-two-numbers/
// O(max(n, m)) time complexity
// O(max(n, m)) space complexity, where
// n and m are lengths of the first linkedList and second linkedList
function addTwoNumbers(l1, l2) {
  const dummyHead = new ListNode();
  let carry = 0,
    listNode1 = l1,
    listNode2 = l2,
    currentNode = dummyHead;

  while (listNode1 !== null || listNode2 != null) {
    const num1 = listNode1 !== null ? listNode1.val : 0;
    const num2 = listNode2 !== null ? listNode2.val : 0;
    const sum = num1 + num2 + carry;

    carry = Math.floor(sum / 10);
    currentNode.next = new ListNode(sum % 10);
    currentNode = currentNode.next;

    if (listNode1 !== null) listNode1 = listNode1.next;
    if (listNode2 !== null) listNode2 = listNode2.next;
  }

  if (carry !== 0) {
    currentNode.next = new ListNode(carry);
  }

  return dummyHead.next;
}
